relay-server/src/services/health_check.rs

```diff
@@ -138,3 +240,95 @@ impl Service for HealthCheckService {
         });
     }
 }
+
+#[derive(Debug)]
+struct SystemInfo {
+    system: System,
+    last_refresh: Instant,
+    refresh_interval: Duration,
+}
+
+impl SystemInfo {
+    pub fn new(refresh_interval: Duration) -> Self {
+        let system = System::new_with_specifics(
+            RefreshKind::new().with_memory(MemoryRefreshKind::everything()),
+        );
+
+        Self {
+            system,
+            last_refresh: Instant::now(),
+            refresh_interval,
+        }
+    }
+
+    pub fn memory(&mut self) -> Memory {
+        self.refresh();
+
+        // Use the cgroup if available in case Relay is running in a container.
+        if let Some(cgroup) = self.system.cgroup_limits() {
+            Memory {
+                used: cgroup.total_memory.saturating_sub(cgroup.free_memory),
+                total: cgroup.total_memory,
+            }
+        } else {
+            Memory {
+                used: self.system.used_memory(),
+                total: self.system.total_memory(),
+            }
+        }
+    }
+
+    fn refresh(&mut self) {
+        let now = Instant::now();
+        if (now - self.last_refresh) >= self.refresh_interval {
```

Comment:
Could this subtraction lead to a panic? I've checked some alternatives in the [docs](https://doc.rust-lang.org/std/time/struct.Instant.html), but there's a note in all of them that they may reintroduce panics in the future.

For readability I rather `self.last_refresh.elapsed()`, but not strongly opinionated.

Reply:
I forgot about `elapsed`, that is the winner here 👍